File: src/test/java/no/fintlabs/validation/constraints/ValueParsableAsTypeValidatorTest.java
package no.fintlabs.validation.constraints;

import no.fintlabs.model.configuration.dtos.ValueMappingDto;
import no.fintlabs.model.configuration.entities.ValueMapping;
import no.fintlabs.validation.valueparsability.ValueParsabilityValidator;
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.*;

class ValueParsableAsTypeValidatorTest {

    private ValueParsabilityValidator valueParsabilityValidator1;
    private ValueParsabilityValidator valueParsabilityValidator2;
    private ValueParsableAsTypeValidator valueParsableAsTypeValidator;

    @BeforeEach
    void setup() {
        valueParsabilityValidator1 = mock(ValueParsabilityValidator.class);
        valueParsabilityValidator2 = mock(ValueParsabilityValidator.class);
        valueParsableAsTypeValidator = new ValueParsableAsTypeValidator(
                Arrays.asList(valueParsabilityValidator1, valueParsabilityValidator2)
        );
        HibernateConstraintValidatorContext hibernateConstraintValidatorContext = mock(HibernateConstraintValidatorContext.class);
        when(hibernateConstraintValidatorContext.addMessageParameter(any(String.class), any())).thenReturn(hibernateConstraintValidatorContext);
    }

    @Test
    void shouldReturnTrueIfAllFieldParsabilityValidatorsReturnTrue() {
        when(valueParsabilityValidator1.isValid(any(ValueMappingDto.class))).thenReturn(true);
        when(valueParsabilityValidator2.isValid(any(ValueMappingDto.class))).thenReturn(true);

        boolean valid = valueParsableAsTypeValidator.isValid(
                ValueMappingDto.builder().type(ValueMapping.Type.STRING).build(),
                mock(HibernateConstraintValidatorContext.class)
        );

        assertTrue(valid);
    }

    @Test
    void shouldReturnFalseIfOneFieldParsabilityValidatorReturnsFalse() {
        when(valueParsabilityValidator1.isValid(any(ValueMappingDto.class))).thenReturn(true);
        when(valueParsabilityValidator2.isValid(any(ValueMappingDto.class))).thenReturn(false);

        boolean valid = valueParsableAsTypeValidator.isValid(
                ValueMappingDto.builder().type(ValueMapping.Type.STRING).build(),
                mock(HibernateConstraintValidatorContext.class)
        );

        assertFalse(valid);
    }
}

File: src/test/kotlin/no/novari/flyt/configuration/validation/constraints/ValueParsableAsTypeValidatorTest.kt
package no.novari.flyt.configuration.validation.constraints

import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto
import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import no.novari.flyt.configuration.validation.valueparsability.ValueParsabilityValidator
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class ValueParsableAsTypeValidatorTest {
    private lateinit var valueParsabilityValidator1: ValueParsabilityValidator
    private lateinit var valueParsabilityValidator2: ValueParsabilityValidator
    private lateinit var valueParsableAsTypeValidator: ValueParsableAsTypeValidator
    private lateinit var hibernateContext: HibernateConstraintValidatorContext

    @BeforeEach
    fun setup() {
        valueParsabilityValidator1 = mock()
        valueParsabilityValidator2 = mock()
        valueParsableAsTypeValidator =
            ValueParsableAsTypeValidator(listOf(valueParsabilityValidator1, valueParsabilityValidator2))
        hibernateContext = mock()
        whenever(hibernateContext.addMessageParameter(any(), any())).thenReturn(hibernateContext)
    }

    @Test
    fun shouldReturnTrueIfAllFieldParsabilityValidatorsReturnTrue() {
        whenever(valueParsabilityValidator1.isValid(any<ValueMappingDto>())).thenReturn(true)
        whenever(valueParsabilityValidator2.isValid(any<ValueMappingDto>())).thenReturn(true)

        val valid =
            valueParsableAsTypeValidator.isValid(
                ValueMappingDto.builder().type(ValueMapping.Type.STRING).build(),
                hibernateContext,
            )

        assertTrue(valid)
    }

    @Test
    fun shouldReturnFalseIfOneFieldParsabilityValidatorReturnsFalse() {
        whenever(valueParsabilityValidator1.isValid(any<ValueMappingDto>())).thenReturn(true)
        whenever(valueParsabilityValidator2.isValid(any<ValueMappingDto>())).thenReturn(false)

        val valid =
            valueParsableAsTypeValidator.isValid(
                ValueMappingDto.builder().type(ValueMapping.Type.STRING).build(),
                hibernateContext,
            )

        assertFalse(valid)
    }
}
